### PR TITLE
feat: keyboardevent

### PR DESCRIPTION
**Pull request by cptbtptpbcptdtptp**
link: [Design](https://github.com/oasis-engine/engine/wiki/Keyboard-Input-design)

### CODE REVIEW

File: packages/core/src/input/keyboard/KeyCode.ts [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    OSRight,
+    OSLeft,
+    AudioVolumeUp,
+    AudioVolumeDown,
+    Undo,
+    Copy,
+    Paste,
+    Cut,
+    Help
+}
```
Comment (GuoLei1990) [inline packages/core/src/input/keyboard/KeyCode.ts:155]: Should add comments for enum items

File: packages/core/src/Script.ts [hunk 2]
```diff
@@ -117,6 +124,18 @@ export class Script extends Component {
...
+   * Called when the keyboard is pressed and don't save the evt.
+   * @param evt - Dispatched key event , only valid in the current frame
+   */
+  onKeyDown(evt: KeyEvent): void {}
+
+  /**
+   * Called when the keyboard is up.
+   * @param evt - Dispatched key event
+   */
+  onKeyUp(evt: KeyEvent): void {}
```
Comment (GuoLei1990) [inline packages/core/src/Script.ts:137]: `KeyDown` and `KeyUp` is global behavior, not entity. Script do not need this

File: packages/core/src/input/keyboard/KeyCode.ts (hunk 1 above)
Comment (cptbtptpbcptdtptp) [reply packages/core/src/input/keyboard/KeyCode.ts:207]: Done.

File: packages/core/src/input/enums/KeyCode.ts [hunk 3]
```diff
@@ -0,0 +1,424 @@
+/**
+ * The number of the code in the keyboard event.
+ * Keep up with W3C standards.
+ * https://www.w3.org/TR/2017/CR-uievents-code-20170601/
+ */
+export enum KeyCode {
```
Comment (GuoLei1990) [inline packages/core/src/input/enums/KeyCode.ts:6]: use `Key` is better than `KeyCode` to understand for developers.

File: packages/core/src/input/enums/KeyCode.ts [hunk 4]
```diff
@@ -0,0 +1,424 @@
+/**
+ * The number of the code in the keyboard event.
```
Comment (GuoLei1990) [inline packages/core/src/input/enums/KeyCode.ts:2]: Don't need to express concepts of 'number, for example `Defines the keys on a keyboard.` is enough.

File: packages/core/src/input/enums/KeyCode.ts [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+    NumpadParenRight,
+    /**
+     * * on a phone or remote control device. This key is typically found below the 7 key and to the left of the 0 key.
+     * Use "NumpadMultiply" for the * key on numeric keypads.
+     */
+    NumpadStar,
+    /** - */
+    NumpadSubtract,
+
+    /** List of code values for keys in the Function section. */
```
Comment (GuoLei1990) [inline packages/core/src/input/enums/KeyCode.ts:289]: Delete this line comments.

File: packages/core/src/input/enums/KeyCode.ts [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+    /** ↓ */
+    ArrowDown,
+    /** ← */
+    ArrowLeft,
+    /** → */
+    ArrowRight,
+    /** ↑ */
+    ArrowUp,
+
+    /** List of code values for keys in the Numpad section. */
```
Comment (GuoLei1990) [inline packages/core/src/input/enums/KeyCode.ts:189]: Delete this line comments.

File: packages/core/src/input/InputManager.ts [hunk 7]
```diff
@@ -37,6 +79,10 @@ export class InputManager {
     if (canvas instanceof HTMLCanvasElement) {
       this._enabled = true;
       this._pointerManager = new PointerManager(engine, canvas);
+      this._keyboardManager = new KeyboardManager();
+      window.addEventListener('blur', () => {
```
Comment (GuoLei1990) [inline packages/core/src/input/InputManager.ts:83]: Why this event register is not in `KeyboardManager`?

File: packages/core/src/input/keyboard/KeyboardManager.ts [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+                        }
+                        break;
+                    case 'keyup':
+                        const delIndex = _curHeldDownKeyToIndexMap[codeNumber];
+                        if (delIndex) {
+                            _curHeldDownKeyToIndexMap[codeNumber] = null;
+                            const swapCode = _curFrameHeldDownList.deleteByIndex(delIndex - 1);
+                            swapCode && (_curHeldDownKeyToIndexMap[swapCode] = delIndex);
+                        }
+                        _curFrameUpList.add(codeNumber);
```
Comment (GuoLei1990) [inline packages/core/src/input/keyboard/KeyboardManager.ts:69]: Maybe add mant times with the same key

**Review by singlecoder — APPROVED** [review]
+1

File: packages/core/src/input/enums/Keys.ts [hunk 9]
```diff
@@ -0,0 +1,410 @@
+/**
+ * The keys of the keyboard.
+ * Keep up with W3C standards.(https://www.w3.org/TR/2017/CR-uievents-code-20170601/)
+ */
+export enum Keys {
+    /** `~ on a US keyboard. This is the 半角/全角/漢字 (hankaku/zenkaku/kanji) key on Japanese keyboards. */
+    Backquote,
```
Comment (GuoLei1990) [inline packages/core/src/input/enums/Keys.ts:7]: Maybe this should be `BackQuote`, W3C is not Right

**Review by GuoLei1990 — APPROVED** [review]
+1

**Review by singlecoder — APPROVED** [review]
+1

**Pull request merged**